// fumadocs-ui/dist/components/callout.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { CircleCheck, CircleX, Info, TriangleAlert } from '../icons.js';
import { forwardRef } from 'react';
import { cn } from '../utils/cn.js';
import { cva } from 'class-variance-authority';
const calloutVariants = cva('my-4 flex gap-2 rounded-lg border border-s-2 bg-fd-card p-3 text-sm text-fd-card-foreground shadow-md', {
    variants: {
        type: {
            info: 'border-s-blue-500/50',
            warn: 'border-s-orange-500/50',
            error: 'border-s-red-500/50',
            success: 'border-s-green-500/50',
        },
    },
});
export const Callout = forwardRef(({ className, children, title, type = 'info', icon, ...props }, ref) => {
    if (type === 'warning')
        type = 'warn';
    return (_jsxs("div", { ref: ref, className: cn(calloutVariants({
            type: type,
        }), className), ...props, children: [icon ??
                {
                    info: _jsx(Info, { className: "size-5 fill-blue-500 text-fd-card" }),
                    warn: (_jsx(TriangleAlert, { className: "size-5 fill-orange-500 text-fd-card" })),
                    error: _jsx(CircleX, { className: "size-5 fill-red-500 text-fd-card" }),
                    success: (_jsx(CircleCheck, { className: "size-5 fill-green-500 text-fd-card" })),
                }[type], _jsxs("div", { className: "min-w-0 flex flex-col gap-2 flex-1", children: [title ? _jsx("p", { className: "font-medium !my-0", children: title }) : null, _jsx("div", { className: "text-fd-muted-foreground prose-no-margin empty:hidden", children: children })] })] }));
});
Callout.displayName = 'Callout';
