// fumadocs-core/dist/highlight/client.js
"use client";
import {
  _highlight,
  _renderHighlight,
  highlight
} from "../chunk-KNWSJ4IF.js";
import "../chunk-MLKGABMK.js";

// src/highlight/client.tsx
import {
  useId,
  useMemo,
  useRef,
  use,
  useState,
  useLayoutEffect
} from "react";
import { jsx } from "react/jsx-runtime";
var jsEngine;
function getHighlightOptions(from) {
  if (from.engine) return from;
  if (!jsEngine) {
    jsEngine = import("shiki/engine/javascript").then(
      (res) => res.createJavaScriptRegexEngine()
    );
  }
  return {
    ...from,
    engine: jsEngine
  };
}
function useShiki(code, {
  defaultValue,
  withPrerenderScript = false,
  ...options
}, deps) {
  const markupId = useId();
  const key = useMemo(
    () => deps ? JSON.stringify(deps) : `${options.lang}:${code}`,
    [code, deps, options.lang]
  );
  const shikiOptions = getHighlightOptions(options);
  const currentTask = useRef({
    key,
    aborted: false
  });
  const [rendered, setRendered] = useState(() => {
    if (defaultValue) return defaultValue;
    const element = withPrerenderScript && typeof document !== "undefined" ? document.querySelector(`[data-markup-id="${markupId}"]`) : null;
    const attr = element?.getAttribute("data-markup");
    if (attr) {
      const hast = JSON.parse(attr);
      return renderHighlightWithMarkup(markupId, hast, shikiOptions, attr);
    }
    currentTask.current = void 0;
    const Pre = options.components?.pre ?? "pre";
    const Code = options.components?.code ?? "code";
    return /* @__PURE__ */ jsx(Pre, { children: /* @__PURE__ */ jsx(Code, { children: code }) });
  });
  useLayoutEffect(() => {
    if (currentTask.current?.key === key) return;
    if (currentTask.current) {
      currentTask.current.aborted = true;
    }
    const task = {
      key,
      aborted: false
    };
    currentTask.current = task;
    void highlight(code, shikiOptions).then((result) => {
      if (!task.aborted) setRendered(result);
    });
  }, [key]);
  if (typeof window === "undefined") {
    return use(
      _highlight(code, shikiOptions).then(
        (tree) => renderHighlightWithMarkup(markupId, tree, shikiOptions)
      )
    );
  }
  return rendered;
}
function renderHighlightWithMarkup(id, tree, shikiOptions, rawAttr) {
  const Pre = shikiOptions.components?.pre ?? "pre";
  return _renderHighlight(tree, {
    ...shikiOptions,
    components: {
      ...shikiOptions.components,
      pre: (props) => /* @__PURE__ */ jsx(
        Pre,
        {
          ...props,
          "data-markup-id": id,
          "data-markup": rawAttr ?? JSON.stringify(tree)
        }
      )
    }
  });
}
export {
  useShiki
};
