// fumadocs-mdx/dist/chunk-4LSNX4UE.js
// src/runtime/index.ts
import fs from "node:fs";
var _runtime = {
  doc(files) {
    return files.map((file) => {
      const { default: body, frontmatter, ...exports } = file.data;
      let cachedContent;
      return {
        body,
        ...exports,
        ...frontmatter,
        get content() {
          cachedContent ??= fs.readFileSync(file.info.absolutePath).toString();
          return cachedContent;
        },
        _exports: file.data,
        _file: file.info
      };
    });
  },
  meta(files) {
    return files.map((file) => {
      return {
        ...file.data,
        _file: file.info
      };
    });
  },
  docs(docs, metas) {
    const parsedDocs = this.doc(docs);
    const parsedMetas = this.meta(metas);
    return {
      docs: parsedDocs,
      meta: parsedMetas,
      toFumadocsSource() {
        return createMDXSource(parsedDocs, parsedMetas);
      }
    };
  }
};
function createMDXSource(docs, meta = []) {
  return {
    files: () => resolveFiles({
      docs,
      meta
    })
  };
}
function resolveFiles({ docs, meta }) {
  const outputs = [];
  for (const entry of docs) {
    outputs.push({
      type: "page",
      path: entry._file.path,
      data: entry
    });
  }
  for (const entry of meta) {
    outputs.push({
      type: "meta",
      path: entry._file.path,
      data: entry
    });
  }
  return outputs;
}

export {
  _runtime,
  createMDXSource,
  resolveFiles
};
